### ISSUE TITLE
dnssec: GETDNS_EXTENSION_TRUE

### ISSUE THREAD

**Issue by CarlTuckerson** [getdnsapi/stubby#184]
Installed on Windows 10 pro using the Cloudflare severs.

Not sure what dnssec: GETDNS_EXTENSION_TRUE is doing.

Can't reach http://www.dnssec-bogus.sg/ whether it's enabled or commented out as it is by default. Can reach the page when not using Stubby and Cloudflare DNS.

The test at http://dnssec.vs.uni-due.de/ shows "Yes, your DNS resolver validates DNSSEC signatures" either way.

dig sigfail.verteiltesysteme.net @127.0.0.1 shows status: SERVFAIL, no A record returned, whether this option is commented out or not.

https://rootcanary.org/test.html returns the same results either way.

The only thing that shows a difference is the results at https://1.1.1.1/help. With the option commented out, it shows Yes for both Connected to 1.1.1.1 and Using DNS over TLS (DoT), as it also does from an Android tablet using Cloudflare's app, and from a phone using Cloudflare via Android Pie's option for DoT. With the option enabled in the stubby config on Windows, Cloudflare returns a No for both items.

Well, there is one other place I see a difference: it does show dnssec off when commented out, and on when enabled, when stubby is started at the command line.

**Comment by CarlTuckerson** [comment]
I now have DNSSEC validation enabled in Stubby on Windows 10 AND pass Cloudflare's tests in the browser (Edge in my case).
I thought simply enabling the line that was commented out in stubby.yml was sufficient:
dnssec: GETDNS_EXTENSION_TRUE
Then, per the info at https://dnsprivacy.org/wiki/display/DP/Configuring+Stubby I added the following:
dnssec_return_status: GETDNS_EXTENSION_TRUE
Still, with both those lines, https://1.1.1.1/help showed not connected to 1.1.1.1 and not using DoT.
Likewise, https://www.cloudflare.com/ssl/encrypted-sni/ showed not using secure DNS.
Adding the following line corrects the problem:
return_both_v4_and_v6: GETDNS_EXTENSION_TRUE
With it, both Cloudflare pages return the expected results. Comment it out, and it fails.

**Comment by saradickinson** [comment]
Q: Not sure what `dnssec: GETDNS_EXTENSION_TRUE` is doing.
A: This flag enables DNSSEC validation in the client (Stubby) on your local machine as opposed to relying on the DNSSEC validation done by the remote resolver (Cloudflare). In either case not being able to reach  http://www.dnssec-bogus.sg/  is correct behaviour because the validation fails either locally or in the resolver.  If you want to be sure you have local validation working then you would need to point stubby at a resolver that doesn't do DNSSEC validation e.g. 9.9.9.10 and then test access to the sites you mention.

Q: Then, per the info at https://dnsprivacy.org/wiki/display/DP/Configuring+Stubby I added the following: `dnssec_return_status: GETDNS_EXTENSION_TRUE`
You don't need this if you have `dnssec: GETDNS_EXTENSION_TRUE`, it just returns additional information that Stubby doesn't use.

Q: The only thing that shows a difference is the results at https://1.1.1.1/help.
A: I don't know how this page actually works but it might just show the Cloudflare servers view of your connection or do probe queries, I don't think it goes digging around your system DNS settings or monitors your outgoing traffic. But I believe it is reporting incorrectly for some reason when  `dnssec: GETDNS_EXTENSION_TRUE`  is enabled. I can see traffic going to 1.1.1.1 and only 1.1.1.1 yet the tool reports no connection as you say.... I suspect the tool depends on IPv6 connectivity in some cases - I would raise an issue with Cloudflare as stubby is behaving correctly.  Be aware that most online connection checkers assume that ONLY the resolver does DNSSEC validation and won't/can't test properly for local validation.

**Comment by CarlTuckerson** [comment]
Thanks, pointing to 9.9.9.10 instead of cloudflare does prove all the expected results when only dnssec: is enabled or disabled in the config file.

The stubby.yml comments section is correct. However, on the page for Windows installer for Stubby, no mention of DNSSEC is found there, except for a comment linking to the Configuring Stubby page, where dnssec_return_status: is said to be the option to turn on DNSSEC. I enabled both since it didn't occur to me that the several tests I looked at would detect the stub resolver's DNSSEC validation even when it connected to a non-DNSSEC server. Now I can see that the dnssec: option is sufficient.

The Cloudflare pages failing to detect TLS encryption is fixed with the return_both_v4_and_v6: option, though (1.1.1.1/help and www.cloudflare.com/ssl/encrypted-sni/). So, everything working fine, now. Thanks.
